### ISSUE TITLE
RetryObjectReadSource::Read() doesn't retry hard enough

### ISSUE THREAD

**Issue by dopiera** [googleapis/google-cloud-cpp#2964]
TL;DR; two consecutive `recv()` failures are enough to to exhaust any retry policy on `Client::ReadObject()`.

`Client::ReadObject()` translates into `RetryObjectReadSource::Read()`, calls `CurlDownloadRequest::Read()`.

`RetryObjectReadSource::Read()` does  the following, when `CurlDownloadRequest::Read()` fails:
- create a new `CurlDownloadRequest` with retries (fail if neither retry is successful)
- retry `Read()` on the newly recreated `CurlDownloadRequest` *only once*

According to `strace` and code inspection, `CurlDownloadRequest` can succeed if there is no network at all, so any communication is actually tried only on `CurlDownloadRequest::Read()`, which is not retried.

That way, a transient network failure makes `RetryObjectReadSource::Read()` ignore the retry policy and propagates the error straight away.

**Comment by dopiera** [comment]
FWIW, it looks like the `curl` CLI tool doesn't reinitialize the curl handle - it just keeps running `curl_easy_perform()` - perhaps we should do the same?

**Comment by coryan** [comment]
@dopiera I think this is a duplicate of #2890.  I am going to close it, let me know if you disagree.
